File: 프로그래머스/3/1832. 보행자 천국/보행자 천국.java
import java.util.LinkedList;
import java.util.Queue;

public class Solution {

    static class Point {

        int fromTop;
        int fromLeft;
        int type;

        public Point(int type) {
            this.type = type;
        }
    }

    static int MOD = 20170805;

    public static final int[] dx = {1, 0}; // 우, 하
    public static final int[] dy = {0, 1};
    public static Point[][] graph;

    public static int solution(int m, int n, int[][] cityMap) {
        int answer = 0;
        graph = new Point[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                graph[i][j] = new Point(cityMap[i][j]);
            }
        }

        for (int i = 1; i < m; i++) {
            if (cityMap[i][0] == 1) {
                break;
            } else {
                graph[i][0].fromTop = 1;
            }
        }

        for (int i = 1; i < n; i++) {
            if (cityMap[0][i] == 1) {
                break;
            } else {
                graph[0][i].fromLeft = 1;
            }
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (graph[i][j].type == 1) {
                    continue;
                }

                graph[i][j].fromLeft = graph[i][j - 1].fromLeft;
                if (graph[i][j - 1].type != 2) {
                    graph[i][j].fromLeft = (graph[i][j].fromLeft + graph[i][j - 1].fromTop) % MOD;
                }

                graph[i][j].fromTop = graph[i - 1][j].fromTop;
                if (graph[i - 1][j].type != 2) {
                    graph[i][j].fromTop = (graph[i][j].fromTop + graph[i - 1][j].fromLeft) % MOD;
                }
            }
        }

        answer = (graph[m - 1][n - 1].fromLeft + graph[m - 1][n - 1].fromTop) % MOD;

        return answer;
    }
}

File: 프로그래머스/3/1832. 보행자 천국/보행자 천국.cpp
#include <vector>

using namespace std;

int MOD = 20170805;

int GetLeftCase(int startI, int startJ, vector<vector<int>>& map, vector<vector<int>>& dp)
{
    for(int j = startJ; j >= 0; --j)
    {
        if (map[startI][j] == 2)
            continue;
        
        return dp[startI][j];
    }
    
    return 0;
}

int GetTopCase(int startI, int startJ, vector<vector<int>>& map, vector<vector<int>>& dp)
{
    for(int i = startI; i >= 0; --i)
    {
        if (map[i][startJ] == 2)
            continue;
        
        return dp[i][startJ];
    }
    
    return 0;
}

// 전역 변수를 정의할 경우 함수 내에 초기화 코드를 꼭 작성해주세요.
int solution(int m, int n, vector<vector<int>> city_map) {
    int answer = 0;
    
    vector<vector<int>> dp (m, vector<int> (n, 0));
    dp[0][0] = 1;
    
    for(int i = 0; i < m; ++i)
    {
        for(int j = 0; j < n; ++j)
        {
            if (i == 0 && j == 0)
                continue;
            
            if (city_map[i][j] == 1)
            {
                dp[i][j] = 0;
                continue;
            }
            
            dp[i][j] = (GetLeftCase(i, j - 1, city_map, dp) + GetTopCase(i - 1, j, city_map, dp)) % MOD;
        }
    }
    
    answer = dp.back().back() % MOD;
    return answer;
}